File: scripts/utilities/check-for-pretuned-sizes.py
#!/usr/bin/python

"""Copyright (C) 2016-2022 Advanced Micro Devices, Inc. All rights reserved.

   Permission is hereby granted, free of charge, to any person obtaining a copy
   of this software and associated documentation files (the "Software"), to deal
   in the Software without restriction, including without limitation the rights
   to use, copy, modify, merge, publish, distribute, sublicense, and/or sell cop-
   ies of the Software, and to permit persons to whom the Software is furnished
   to do so, subject to the following conditions:

   The above copyright notice and this permission notice shall be included in all
   copies or substantial portions of the Software.

   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IM-
   PLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
   FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
   COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
   IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNE-
   CTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
"""

import yaml
import sys
import os
import subprocess
from subprocess import PIPE
from tempfile import mkdtemp
from shutil import rmtree

#
minimumVersionIdx       = 0
prefixIdx               = 1
architectureIdx         = 2
deviceNamesIdx          = 3
problemTypeStateIdx     = 4
solutionListIdx         = 5
indexOrderIdx           = 6
exactLogicListIdx       = 7       #[[m, n, batch_size, k, ...], [kernelIdx, GFLOPS]]
mIdx        = 0
nIdx        = 1
batchIdx    = 2
kIdx        = 3
rangeLogicIdx           = 8
tileSelectionLogicIdx   = 9       #Only if tile aware selection is enabled

helpMessage = """Script for determining which problem sizes in a log file have been pre-tuned for
by Tensile.
Usage:
python3 check-for-pretuned-sizes.py -f logfilePath
                                      [-c commit] [-a architecture] [-u] [-l] [--help]
Options:
-f       Log file containing benchmark tests.
-a       Architecture. Matches any problem containing this argument.
         ex: vega10
-c       Specify commit of rocBLAS to check. If unspecified, will use most recent master.
-l       Attempt to detect the installed version of rocBLAS
         (Currently just works when installing from cloned repository)
-u       Compares detected problem sizes to those found in the most recent develop commit.
--help   Displays this message.
NOTE: Flags cannot currently be combined together as in -lu, they must be specified
separately in the form -l -u
"""

usageMessage = """Usage:
python3 check-for-pretuned-sizes.py -f logfilePath
                                      [-c commit] [-a architecture] [-u] [-l] [--help]"""


def removeChar(text, removedChar):
    return "".join(filter(lambda c : c != removedChar, text))

class ParseOptionError(Exception):
    pass

#Same usage as getopt but doesn't fail on unknown args and returns dict
def parseOptions(textList, shortArgs, longArgs=[]):
    hasArg = []
    lastChar = ':'
    optionList = []
    #Set keys for short arguments
    for c in shortArgs:
        #':' indicates c takes an argument
        if c == ':':
            if not lastChar.isalpha():
                raise ParseOptionError("Error: Colon must follow alphabetic character")
            hasArg[-1] = True
        else:
            optionList.append('-' + c)
            hasArg.append(False)
        lastChar = c

    #Set keys for long arguments
    for s in longArgs:
        #'=' indicates s takes an argument
        if s[-1] == '=':
            hasArg.append(True)
            optionList.append("--" + s[:-1])
        else:
            hasArg.append(False)
            optionList.append("--" + s)

    hasArg = dict(zip(optionList,
        hasArg))

    #Extract options from textList
    optionDict = dict([])
    needsArg = None
    for item in textList:
        #If this is an argument of a previous option...
        if needsArg != None:
            #Store associated with it
            optionDict[needsArg] = item
            needsArg = None
        #Otherwise check if it's a valid option
        elif item in optionList:
            optionDict[item] = None     #Set as none for now
            if hasArg[item]:
                needsArg = item         #Set to capture next item

    return dict(zip(
        list(map(lambda s : removeChar(s, '-'), optionDict.keys())),
        optionDict.values()))

#assumes b has at least as many periods as a
def isVersionGreaterThanOrEqual(versionA, versionB):
    for a, b in zip(versionA.split('.'), versionB.split('.')):
        if int(a) != int(b):
            return int(a) > int(b)
    return True     #Is equal


def shellCmd(command):
    spReturn = subprocess.run(command, shell=True, stdout=PIPE, stderr=PIPE)
    return spReturn.stdout.decode('utf-8').rstrip('\n')


def getInstalledRocBLASCommitHash():
    versionString = shellCmd("dpkg -s rocblas | grep Version") #Format "Version X.XX.X.XXXX-CommitHash"
    spaceSplit = versionString.split()
    if len(spaceSplit) > 1:
        dashSplit = spaceSplit[1].split('-')
        if len(dashSplit) > 1:
            return dashSplit[1]
        else:
            print("Error: No commit hash found in \"dpkg -s rocblas | grep Version\" command")
            return ""
    else:
        print("Error: Could not parse output from \"dpkg -s rocblas | grep Version\" command")
        return ""


def cloneRepository(destinationPath):
    print("Cloning repository...")
    shellCmd("git clone https://github.com/ROCmSoftwarePlatform/rocBLAS.git %s" % destinationPath)

def checkoutSpecifiedCommit(destinationPath, commit):
    originalPath = shellCmd("pwd").rstrip('\n')
    os.chdir(destinationPath)
    shellCmd("git checkout %s" % commit)
    os.chdir(originalPath)

def checkoutInstalledBranch(destinationPath):
    rocBLASCommitHash = getInstalledRocBLASCommitHash()
    checkoutSpecifiedCommit(destinationPath, rocBLASCommitHash)

def checkoutMostRecentBranch(destinationPath, branchname):
    originalPath = shellCmd("pwd").rstrip('\n')
    os.chdir(destinationPath)
    shellCmd("git checkout %s" % branchname)
    os.chdir(originalPath)

def convertToExplicitType(t):
    typeMap = {
        'h': 'f16_r',
        's': 'f32_r',
        'd': 'f64_r',
        'c': 'f32_r',
        'z': 'f64_r'}
    if t in typeMap:
        return typeMap[t]
    else:
        return t

def convertArgumentTypesToKernelIdentifier(aType, bType, cType, dType, computeType):
    aType       = convertToExplicitType(aType)
    bType       = convertToExplicitType(bType)
    cType       = convertToExplicitType(cType)
    dType       = convertToExplicitType(dType)
    computeType = convertToExplicitType(computeType)

    argumentsToRocblasType = {
        ('f16_r',  'f16_r',  'f16_r',  'f16_r',  'f16_r') : 'half_precision',
        ('f16_r',  'f16_r',  'f16_r',  'f16_r',  'f32_r') : 'hpa_half_precision',
        ('f32_r',  'f32_r',  'f32_r',  'f32_r',  'f32_r') : 'single_precision',
        ('f64_r',  'f64_r',  'f64_r',  'f64_r',  'f64_r') : 'double_precision',
        ('i8_r',   'i8_r',   'i8_r',   'i8_r',   'i32_r') : 'int8_precision',
        ('bf16_r', 'bf16_r', 'bf16_r', 'bf16_r', 'bf16_r'): 'bf16_precision',
        ('bf16_r', 'bf16_r', 'bf16_r', 'bf16_r', 'f32_r') : 'hpa_bf16_precision',
        ('f16_c',  'f16_c',  'f16_c',  'f16_c',  'f16_c') : 'half_precision_complex',
        ('f16_c',  'f16_c',  'f16_c',  'f16_c',  'f32_c') : 'hpa_half_precision_complex',
        ('f32_c',  'f32_c',  'f32_c',  'f32_c',  'f32_c') : 'single_precision_complex',
        ('f64_c',  'f64_c',  'f64_c',  'f64_c',  'f64_c') : 'double_precision_complex',
        ('i8_c',   'i8_c',   'i8_c',   'i8_c',   'i32_c') : 'int8_precision_complex',
        ('bf16_c', 'bf16_c', 'bf16_c', 'bf16_c', 'bf16_c'): 'bf16_precision_complex',
        ('bf16_c', 'bf16_c', 'bf16_c', 'bf16_c', 'f32_c') : 'hpa_bf16_precision_complex'}
    rocblasTypeToKernelIdentifier = {
        'half_precision'          : 'HB.',
        'hpa_half_precision'      : 'HBH.',
        'single_precision'        : 'SB.',
        'double_precision'        : 'DB.',
        'int8_precision'          : '4xi8BH.',
        'bf16_precision'          : 'BB.',
        'hpa_bf16_precision'      : 'BBH.',
        'single_precision_complex': 'CB.',
        'double_precision_complex': 'ZB.'}

    try:
        rocblasType = argumentsToRocblasType[(aType, bType, cType, dType, computeType)]
        kernelIdentifier = rocblasTypeToKernelIdentifier[rocblasType]
    except KeyError:
        print("Error: Unrecognized argument type combination (a_type %s b_type %s c_type %s d_type %s compute_type %s)" %(aType, bType, cType, dType, computeType))
        return 'NOT VALID'
    return kernelIdentifier
    #Comes from the roblas_common.yaml file

#Gets one of the first or second keys provided
def getOne(d, key1, key2):
    if key1 in d:
        return d[key1]
    else:
        return d[key2]

def supportedProblemType(problemType):
    return 'gemm' in problemType


#Bind as member variable to dict

class ProblemDescription:
    def __init__(self, benchmarkText):
        self.m = 1
        self.n = 1
        self.k = 1
        self.batch_count = 1

        try:
            optDict= parseOptions(
                benchmarkText.split(),
                "m:n:k:f:r:",
                [
                    "batch_count=", "transposeA=", "transposeB=", \
                    "a_type=", "b_type=", "c_type=", "d_type=", "compute_type=",\
                    "precision=", "sizem=", "sizen=", "sizek="\
                ]
            )
            self.gemmType     =     getOne(optDict, 'f','function')
            if not supportedProblemType(self.gemmType):
                return
            self.m            = int(getOne(optDict, 'm', 'sizem'))
            self.n            = int(getOne(optDict, 'n', 'sizen'))
            self.k            = int(getOne(optDict, 'k', 'sizek'))
            transposeA        =            optDict['transposeA'] == 'T'
            transposeB        =            optDict['transposeB'] == 'T'
            if 'ex' in self.gemmType:
                aType         =            optDict['a_type']
                bType         =            optDict['b_type']
                cType         =            optDict['c_type']
                dType         =            optDict['d_type']
                computeType   =            optDict['compute_type']
            else:
                aType         =\
                bType         =\
                cType         =\
                dType         =\
                computeType   =     getOne(optDict, 'r', 'precision')
            self.kernel_flags = convertArgumentTypesToKernelIdentifier(
                aType, bType, cType, dType, computeType)

        except ValueError:
            print("Error: Problem description parameters have invalid type")
        except ParseOptionError:
            print("Error: Input arguments ill formed")

        if 'batch_count' in optDict:
            self.batch_count = int(optDict['batch_count'])

        self.matrix_A = 'Alik' if transposeA else 'Ailk'
        self.matrix_B = 'Bjlk' if transposeB else 'Bljk'

    def __str__(self):
        return "(m=%d n=%d k=%d batch_count=%d transpose_A=%r transpose_B=%r kernel_flags=%s)" \
        % (self.m, self.n, self.k, self.batch_count, self.matrix_A == 'Alik', self.matrix_B == 'Bjlk', self.kernel_flags)

def loadBenchmarkDescriptions(logfilePath):
    try:
        f = open(logfilePath)
    except OSError:
        print("%s file not found." % logfilePath)
        return []

    lines = f.read().split('\n')
    benchmarkList = []

    for line in lines:
        pieces = line.split(maxsplit=1)
        if pieces and "rocblas-bench" in pieces[0]:
            try:
                problem = ProblemDescription(line)
            except KeyError:
                raise KeyError(line)
            if supportedProblemType(problem.gemmType):
                benchmarkList.append(problem)
    return benchmarkList


def matchBetween(matrixFormSet, filename):
    for matrixForm in matrixFormSet:
        if matrixForm[0] in filename and matrixForm[1] in filename and matrixForm[2] in filename:
            return True
    return False

#Search through library logic files in directory to find a matching kernel
def findMatchingKernel(benchDescriptions, architecture, directoryPath):
    #Whether a benchmark description has found a match
    print("Searching %s for logic files..." % directoryPath)
    foundMatch = [False for x in range(len(benchDescriptions))]

    problemSet = set()
    for bench in benchDescriptions:
        problemSet.add((bench.matrix_A, bench.matrix_B, bench.kernel_flags))

    #Filter files by problem types in benchmarks
    fileList = []
    for filename in os.listdir(directoryPath):
        if filename.endswith(".yaml") \
        and architecture in filename \
        and matchBetween(problemSet, filename):
            fileList.append(filename)

    success = False
    #Find a kernel with a matching problem size in the file list
    for filename in fileList:
        foundMatchingKernel = False
        f = open(directoryPath + filename, 'r')
        if f.closed:
            sys.exit("File %s could not be opened" % directoryPath + filename)
        libraryLogic = yaml.safe_load(f)
        f.close()
        logicList = libraryLogic[exactLogicListIdx]
        for i in range(len(benchDescriptions)):
            bench = benchDescriptions[i]
            if bench.matrix_A in filename and bench.matrix_B in filename and bench.kernel_flags in filename:
                #Iterate through kernels to find one with matching size
                for [problemInfo, winningKernelInfo] in logicList:
                    foundMatchingKernel = \
                        problemInfo[mIdx]     == bench.m and \
                        problemInfo[nIdx]     == bench.n and \
                        problemInfo[kIdx]     == bench.k and \
                        problemInfo[batchIdx] == bench.batch_count
                    if foundMatchingKernel: break

                if foundMatchingKernel:
                    print("Match found for %s in file %s" % \
                        (                  bench,      filename))
                    foundMatch[i] = True
    return foundMatch

def findBenchmarkInFile(problemDescriptions):
    pass


def main(argv):
    try:
        optdict = parseOptions(argv, "f:a:c:ul", ["help"])
        if 'help' in optdict.keys():
            print(helpMessage)
            sys.exit()
    except KeyError:
        sys.exit(usageMessage)
    except ParseOptionError:
        sys.exit(usageMessage)

    logpath = ""
    if 'f' in optdict:
        logpath = optdict['f']
    else:
        sys.exit(usageMessage)
    architecture = ""
    if 'a' in optdict:
        architecture = optdict['a']

    directoryPath = mkdtemp()   #"./rocBLASTemp"     #Default path
    libraryPathExtension="/library/src/blas3/Tensile/Logic/asm_full/"
    print("Created temporary directory %s" % directoryPath)
    cloneRepository(directoryPath)

    #Benchmarked problem sizes in log file
    benchDescriptions = loadBenchmarkDescriptions(logpath)
    print("\
-------------------------------------------------------------\n\
-- Finding pre-tuned sizes in specified version of rocblas --\n\
-------------------------------------------------------------")
    if not ('c' in optdict or 'l' in optdict):
        checkoutMostRecentBranch(directoryPath, "master")
    else:
        specifiedCommitHash = optdict['c'] if 'c' in optdict \
                              else getInstalledRocBLASCommitHash()
        checkoutSpecifiedCommit(directoryPath, specifiedCommitHash)
    localMatches = findMatchingKernel(
        benchDescriptions,
        architecture,
        directoryPath+libraryPathExtension)

    if 'u' in optdict:
        print("\
---------------------------------------------------------------\n\
-- Finding pre-tuned sizes in most recent version of rocblas --\n\
---------------------------------------------------------------")
        checkoutMostRecentBranch(directoryPath, "develop")
        recentMatches = findMatchingKernel(
            benchDescriptions,
            architecture,
            directoryPath+libraryPathExtension)

        print("\nMatches in most recent version of rocBLAS but not specified version:")
        for i in range(len(benchDescriptions)):
            if recentMatches[i] and not localMatches[i]:
                print("\t%s" % benchDescriptions[i])


    #Remove temporary directory only if it wasn't provided by user
    print("Removing temporary directory")
    rmtree(directoryPath) #shellCmd("rm -rf %s" % directoryPath)

if __name__=="__main__":
    main(sys.argv[1:])

File: scripts/utilities/check_for_pretuned_sizes_c/check-for-pretuned-sizes.cpp
/* ************************************************************************
 * Copyright (C) 2020-2024 Advanced Micro Devices, Inc. All rights reserved.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell cop-
 * ies of the Software, and to permit persons to whom the Software is furnished
 * to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IM-
 * PLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNE-
 * CTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 *
 * ************************************************************************ */

#include <rocblas/rocblas.h>

#include <algorithm>
#include <cstdlib>
#include <cstring>
#include <fstream>
#include <iostream>
#include <stdexcept>
#include <string>
#include <vector>

// rocBLAS clients contains a substitute for Boost's program_options
#include "../../../clients/benchmarks/program_options.hpp"
using namespace roc;

void init_scalar_value(rocblas_union_t* scalar, rocblas_datatype type, double value)
{
    switch(type)
    {
    case rocblas_datatype_f16_r:
        scalar->h = rocblas_half(value);
        break;
    case rocblas_datatype_f32_r:
        scalar->s = float(value);
        break;
    case rocblas_datatype_f64_r:
        scalar->d = double(value);
        break;
    case rocblas_datatype_i32_r:
        scalar->i = int32_t(value);
        break;
    case rocblas_datatype_f32_c:
        scalar->c = rocblas_float_complex(value);
        break;
    case rocblas_datatype_f64_c:
        scalar->z = rocblas_double_complex(value);
        break;
    default:
        throw std::runtime_error("Unhandled type in init_scalar_value");
    }
}

std::vector<char*> separateIntoTokens(std::string line, char delim = ' ')
{
    std::vector<char*> rv;
    std::string        token;
    for(auto c : line)
    {
        if(c != delim)
            token.push_back(c);
        else if(token.size() > 0)
        {
            rv.push_back(strdup(token.c_str()));
            token.clear();
        }
    }

    if(token.size() > 0)
        rv.push_back(strdup(token.c_str()));

    return rv;
}

//Copied from rocblas_datatype2string.hpp
// clang-format off
inline rocblas_datatype string2rocblas_datatype(const std::string& value)
{
    return
        value == "f16_r" || value == "h" ? rocblas_datatype_f16_r  :
        value == "f32_r" || value == "s" ? rocblas_datatype_f32_r  :
        value == "f64_r" || value == "d" ? rocblas_datatype_f64_r  :
        value == "bf16_r"                ? rocblas_datatype_bf16_r :
        value == "f16_c"                 ? rocblas_datatype_f16_c  :
        value == "f32_c" || value == "c" ? rocblas_datatype_f32_c  :
        value == "f64_c" || value == "z" ? rocblas_datatype_f64_c  :
        value == "bf16_c"                ? rocblas_datatype_bf16_c :
        value == "i8_r"                  ? rocblas_datatype_i8_r   :
        value == "i32_r"                 ? rocblas_datatype_i32_r  :
        value == "i8_c"                  ? rocblas_datatype_i8_c   :
        value == "i32_c"                 ? rocblas_datatype_i32_c  :
        value == "u8_r"                  ? rocblas_datatype_u8_r   :
        value == "u32_r"                 ? rocblas_datatype_u32_r  :
        value == "u8_c"                  ? rocblas_datatype_u8_c   :
        value == "u32_c"                 ? rocblas_datatype_u32_c  :
        static_cast<rocblas_datatype>(-1);
}
// clang-format on

struct Arguments
{
    rocblas_int M;
    rocblas_int N;
    rocblas_int K;

    rocblas_datatype a_type;
    rocblas_datatype b_type;
    rocblas_datatype c_type;
    rocblas_datatype d_type;
    rocblas_datatype compute_type;

    double alpha;
    double beta;

    rocblas_operation transA;
    rocblas_operation transB;

    rocblas_int batch_count;

    char function[64];
};

Arguments parseArguments(char** args, int argCount)
{
    Arguments   arg;
    std::string function;
    std::string precision;
    std::string a_type;
    std::string b_type;
    std::string c_type;
    std::string d_type;
    std::string compute_type;

    char transA;
    char transB;

    options_description desc("rocblas-bench command line options");
    desc.add_options()
        // clang-format off
        ("sizem,m",
         value<rocblas_int>(&arg.M)->default_value(128),
         "Specific matrix size: sizem is only applicable to BLAS-2 & BLAS-3: the number of "
         "rows or columns in matrix.")

        ("sizen,n",
         value<rocblas_int>(&arg.N)->default_value(128),
         "Specific matrix/vector size: BLAS-1: the length of the vector. BLAS-2 & "
         "BLAS-3: the number of rows or columns in matrix")

        ("sizek,k",
         value<rocblas_int>(&arg.K)->default_value(128),
         "Specific matrix size: BLAS-2: the number of sub or super-diagonals of A. BLAS-3: "
         "the number of columns in A and rows in B.")

        ("alpha",
          value<double>(&arg.alpha)->default_value(1.0), "specifies the scalar alpha")

        ("beta",
         value<double>(&arg.beta)->default_value(0.0), "specifies the scalar beta")

        ("function,f",
         value<std::string>(&function),
         "BLAS function to test.")

        ("precision,r",
         value<std::string>(&precision)->default_value("f32_r"), "Precision. "
         "Options: h,s,d,c,z,f16_r,f32_r,f64_r,bf16_r,f32_c,f64_c,i8_r,i32_r")

        ("a_type",
         value<std::string>(&a_type), "Precision of matrix A. "
         "Options: h,s,d,c,z,f16_r,f32_r,f64_r,bf16_r,f32_c,f64_c,i8_r,i32_r")

        ("b_type",
         value<std::string>(&b_type), "Precision of matrix B. "
         "Options: h,s,d,c,z,f16_r,f32_r,f64_r,bf16_r,f32_c,f64_c,i8_r,i32_r")

        ("c_type",
         value<std::string>(&c_type), "Precision of matrix C. "
         "Options: h,s,d,c,z,f16_r,f32_r,f64_r,bf16_r,f32_c,f64_c,i8_r,i32_r")

        ("d_type",
         value<std::string>(&d_type), "Precision of matrix D. "
         "Options: h,s,d,c,z,f16_r,f32_r,f64_r,bf16_r,f32_c,f64_c,i8_r,i32_r")

        ("compute_type",
         value<std::string>(&compute_type), "Precision of computation. "
         "Options: h,s,d,c,z,f16_r,f32_r,f64_r,bf16_r,f32_c,f64_c,i8_r,i32_r")

        ("transposeA",
         value<char>(&transA)->default_value('N'),
         "N = no transpose, T = transpose, C = conjugate transpose")

        ("transposeB",
         value<char>(&transB)->default_value('N'),
         "N = no transpose, T = transpose, C = conjugate transpose")

        //Not currently needed, but may be in the future
        ("batch_count",
         value<rocblas_int>(&arg.batch_count)->default_value(1),
         "Number of matrices. Only applicable to batched and strided_batched routines");

    variables_map vm;
    store(parse_command_line(argCount, args, desc, true), vm);
    notify(vm);

    std::transform(precision.begin(), precision.end(), precision.begin(), ::tolower);
    auto prec = string2rocblas_datatype(precision);
    if(prec == static_cast<rocblas_datatype>(-1))
        throw std::invalid_argument("Invalid value for --precision " + precision);

    arg.a_type = a_type == "" ? prec : string2rocblas_datatype(a_type);
    if(arg.a_type == static_cast<rocblas_datatype>(-1))
        throw std::invalid_argument("Invalid value for --a_type " + a_type);

    arg.b_type = b_type == "" ? prec : string2rocblas_datatype(b_type);
    if(arg.b_type == static_cast<rocblas_datatype>(-1))
        throw std::invalid_argument("Invalid value for --b_type " + b_type);

    arg.c_type = c_type == "" ? prec : string2rocblas_datatype(c_type);
    if(arg.c_type == static_cast<rocblas_datatype>(-1))
        throw std::invalid_argument("Invalid value for --c_type " + c_type);

    arg.d_type = d_type == "" ? prec : string2rocblas_datatype(d_type);
    if(arg.d_type == static_cast<rocblas_datatype>(-1))
        throw std::invalid_argument("Invalid value for --d_type " + d_type);

    arg.compute_type = compute_type == "" ? prec : string2rocblas_datatype(compute_type);
    if(arg.compute_type == static_cast<rocblas_datatype>(-1))
        throw std::invalid_argument("Invalid value for --compute_type " + compute_type);

    snprintf(arg.function, sizeof(arg.function), "%s", function.c_str());

    arg.transA = transA == 'N' ? rocblas_operation_none : rocblas_operation_transpose;
    arg.transB = transB == 'N' ? rocblas_operation_none : rocblas_operation_transpose;

    return arg;
}

int main(int argc, char* argv[])
{
    if(argc == 1 || strcmp(argv[1], "--help") == 0 || strcmp(argv[1], "-h") == 0){
        std::cout << "Parses rocblas-bench log file and determines which gemm problem sizes have been pretuned for in Tensile" << std::endl;
        std::cout << "Usage:\n\tcheck-for-pretuned-sizes <rocblas-bench logfile>" << std::endl;
        return 1;
    }

    rocblas_initialize();
    rocblas_handle handle;
    rocblas_create_handle(&handle);

    std::fstream f{argv[1]};      //Load log file

    char buffer [2048];
    while(f.getline(buffer, 2048)){
        auto tokens = separateIntoTokens(std::string(buffer));
        if(tokens.size() && strstr(tokens[0], "rocblas-bench")){
            auto arg = parseArguments(tokens.data(), tokens.size());
            if(strstr(arg.function, "gemm")){
                rocblas_union_t alpha, beta;
                init_scalar_value(&alpha, arg.compute_type, 1.0);
                init_scalar_value(&beta, arg.compute_type, 1.0);

                // The solution fitness query is initialized to std::numeric_limits<double>::lowest()
                double fitness;
                rocblas_set_solution_fitness_query(handle, &fitness);
                std::cout << "m - " << arg.M << " n - " << arg.N << " k - " << arg.K << std::endl;
                rocblas_gemm_ex(
                    handle,
                    arg.transA,
                    arg.transB,
                    arg.M,
                    arg.N,
                    arg.K,
                    &alpha,
                    nullptr, // A
                    arg.a_type,
                    1, // lda
                    nullptr, // B
                    arg.b_type,
                    1, // ldb
                    &beta,
                    nullptr, // C
                    arg.c_type,
                    1, // ldc
                    nullptr, // D
                    arg.d_type,
                    1, // ldd
                    arg.compute_type,
                    rocblas_gemm_algo_standard,
                    0, // solution_index
                    0 // flags
                );

               if(!fitness)
               {
                   std::cout << "\033[0;32m[Pretuned]\033[0m " << buffer << std::endl;
               }
               else
               {
                   std::cout << "\033[0;31m[Not pretuned]\033[0m " << buffer << std::endl;
               }

               // We reset the solution fitness query to nullptr to avoid a dangling pointer
               rocblas_set_solution_fitness_query(handle, nullptr);
            }
        }

        for (auto* t: tokens)
            free(t);
    }

    f.close();
}
